src/backend/renderer/multigpu/mod.rs

```diff
@@ -831,188 +843,194 @@ where
-                    dma_source: dma_source_ref.take(),
-                    render: RenderDevice::Renderer(render, node),
-                    target: target_ref.take(),
-                    other_renderers: other_renderers_ref.drain(..).collect(),
-                    proxy_framebuffer: std::marker::PhantomData,
-                    log: log.clone(),
-                };
-                let mut frame = MultiFrame {
-                    node,
-                    frame, // we cheat here and use a raw-ptr, because otherwise your associated type would gain an uncostraint lifetime parameter
-                    damage: Vec::new(),
-                    _target: std::marker::PhantomData::<T>,
-                    log,
-                };
+            .render(size, dst_transform)
+            .map_err(Error::Render)?;
 
-                let res = rendering(&mut new_renderer, &mut frame);
+        Ok(MultiFrame {
+            node,
+            frame: Some(frame),
+            render: ptr, // this is fine, as long as we have the frame, this ptr is valid
+            target: &mut self.target,
+            dst_transform,
+            size,
+            damage: Vec::new(),
+            _types: std::marker::PhantomData::<(T, Target)>,
+            log,
+        })
+    }
+}
 
-                // don't return yet, but first reset the references, so we can restore self on error
-                *dma_source_ref = new_renderer.dma_source.take();
-                *target_ref = new_renderer.target.take();
-                *other_renderers_ref = new_renderer.other_renderers.drain(..).collect();
+impl<'a, 'b, 'c, R: GraphicsApi, T: GraphicsApi, Target> MultiFrame<'a, 'b, 'c, R, T, Target>
+where
+    R: 'static,
+    R::Error: 'static,
+    T::Error: 'static,
+    <R::Device as ApiDevice>::Renderer: Offscreen<Target> + ExportDma + ExportMem + ImportDma + ImportMem,
+    <T::Device as ApiDevice>::Renderer: ImportDma + ImportMem,
+    <<R::Device as ApiDevice>::Renderer as Renderer>::Error: 'static,
+    <<T::Device as ApiDevice>::Renderer as Renderer>::Error: 'static,
+{
+    fn finish_internal(&mut self) -> Result<(), Error<R, T>> {
+        if let Some(frame) = self.frame.take() {
+            frame.finish().map_err(Error::Render)?;
 
-                (res, frame.damage)
-            })
-            .map_err(Error::Render);
-
-        // restore self
-        self.dma_source = dma_source;
-        self.target = target;
-        self.other_renderers = other_renderers;
-
-        // then possibly return the error
-        let (res, damage) = res?;
-        let mut damage = damage
-            .into_iter()
-            .map(|rect| {
-                rect.to_logical(1)
-                    .to_buffer(1, dst_transform, &size.to_logical(1))
-            })
-            .collect::<Vec<_>>();
+            // now the frame is gone, lets use our unholy ptr till the end of this call:
+            let render = unsafe { &mut *self.render };
```

Comment:
Properly defending our unsafe use:

```rust
// SAFETY:
// - The renderer will never be invalid because the lifetime of the frame must be shorter than the renderer.
// - The pointer can't be aliased because of the following:
//   - Creating a frame requires an `&mut` reference to the renderer, making the mutable borrow safe.
//   - The mutable reference is used in a function which mutably borrows the frame, that either being `.finish()` (which takes ownership of the frame) or dropping the frame.
```